### ISSUE TITLE
Node goes NotReady during master outage

### ISSUE THREAD

**Issue by dcowden** [kubernetes/kops#3702]
We have what is probably a NOOB question, but we could really use some guidance. We are having a problem in which worker nodes become NotReady when a master (one of three) goes offline.  

We are confused, because with a multi-master system this is not our expecation.

Here are our details:
**kops version**: Version 1.7.0 (git-e04c29d)

**kubectl version**
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T07:00:21Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.9", GitCommit:"19fe91923d584c30bd6db5c5a21e9f0d5f742de8", GitTreeState:"clean", BuildDate:"2017-10-19T16:55:06Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

**cloud provider: aws**

**Cluster setup details**
  * 3 masters and 3 workers, spread across 3 AZs in AWS. 1 worker and 1 master in each AZ
  * loadbalancer type "internal"
  * weave networking ( we tried canal and flannel also, with the same results)
  * cluster name ecom.dev.colinx.com

**Steps to reproduce**
Terminate a master in the AWS console. 

**What happens**
 * As expected, api requests through the api loadbalancer still work.  
 * As expected, the master we killed dies
 * Not expected, the worker in the same AZ as the master we killed goes to 'NotReady', and recovers about 10 minutes after the replaced master comes online.
 * Not expected, helm stops responding during this period, even though the tiller pods are running on the nodes. We think this is because tiller uses api server calls.

**More Info**
In kube-proxy on the node, we observe that the dns name 'api.internal.ecom.dev' is pointed to a list of A records, mapped to each of the masters.  

We suspect that Route53 gives preference to the entry in the same az, and that the kubelet might be 'hanging onto' the old master IP when it is in its death throws. The only problem with this theory is that the TTL on this record is 60 seconds. I would expect the problem to last only 60 seconds, but it seems to take 10 minutes ( pretty consistently, and pretty exactly) for the node to come back-- so if it is simply the DNS thing, we also have a DNS caching problem.

We are somewhat surprised by this configuration-- we'd have guessed that kubelet on the nodes would point to our master load balancer. 

We are also a bit inexperienced about how bad 'NodeNotReady" is.  as noobs, we expect all nodes to be ready except the node being whacked when we're doing an upgrade. Its somewhat scary to see a worker go NotReady. But do we just need to get over that?  

Any help is appreciated!

**Comment by dcowden** [comment]
More information:
Watching the logs of a node that is 'NotReady', we can see that it receives this error message:

`Oct 26 11:04:03 ip-172-21-5-153 kubelet: E1026 11:04:03.588507   13843 kubelet_node_status.go:357] Error updating node status, will retry: error getting node "ip-172-25-51-11.ec2.internal": Get https://api.internal.ecom.dev.colinx.com/api/v1/nodes/ip-172-25-51-11.ec2.internal: net/http: request canceled (Client.Timeout exceeded while awaiting headers)`

However, very oddly, from a shell on the node, the url can be fetched via curl:
```
[root@ip-172-25-51-11 log]# curl -k https://api.internal.ecom.dev.colinx.com
Unauthorized
```
After about 10 minutes, the node recovers

**Comment by georgebuckerfield** [comment]
This doesn't sound normal. Can you see anything in the logs on the affected instance? Or by running `kubectl describe node <affected node>`?

I haven't been able to replicate the worker node showing as NotReady, but I am seeing some stranger behaviour with the internal.api... DNS entry.

I created a multi-master (3) cluster with topology private, load balancer internal and networking calico. After killing two master nodes, one after another (allowing the cluster to recover between), I see the following:

```
root@ip-172-20-61-2:~# dig +short api.internal.kopstest.georgebuckerfield.net 
172.20.70.241
172.20.95.28
172.20.105.106
172.20.114.196
172.20.61.2
```
However, 172.20.95.28 and 172.20.105.106 no longer exist in the cluster:
```
root@ip-172-20-61-2:~# kubectl get node 
NAME                                           STATUS    AGE       VERSION
ip-172-20-111-114.eu-west-1.compute.internal   Ready     54m       v1.7.4
ip-172-20-114-196.eu-west-1.compute.internal   Ready     20m       v1.7.4
ip-172-20-60-205.eu-west-1.compute.internal    Ready     54m       v1.7.4
ip-172-20-61-2.eu-west-1.compute.internal      Ready     56m       v1.7.4
ip-172-20-70-241.eu-west-1.compute.internal    Ready     8m        v1.7.4
```
And you can see in the kube-controller-manager logs them being removed:
```
I1026 16:17:10.755990       5 nodecontroller.go:682] Deleting node (no longer present in cloud provider): ip-172-20-95-28.eu-west-1.compute.internal
```
I've taken a look at a 1.6 cluster we're running in production that was created with kops and this also has an extra IP in the internal API DNS record that no longer matches any instances running in the cluster.

I don't know if this is the same issue affecting you or something different.

**Comment by georgebuckerfield** [comment]
Looks like the DNS issue has been reported here https://github.com/kubernetes/kops/issues/3351

**Comment by dcowden** [comment]
Hi, George, and Thanks!
Yes, we did narrow it down to kube-dns. Since the problem persists for a while, we were able to get an ssh session on the node while it was happening.

From the affected worker using nslookup, dns lookups for the api.internal. name worked fine, but when we instead switched to using the internal kube-dns service ip, lookups timed out.  The node came ready about when kube-dns restored.

I'll have a look at #3551, maybe that will fix our situation.
